* **Objective**:
The main objective of this project is to build a model that  predicts a district’s median housing price. This will be essential to determine whether it is worth investing in a given area. More specifically, our model’s output will be fed to another machine learning system, along with some other signals. So it’s important to make our housing price model as accurate as we can.


* **Current working way**:
Housing prices are currently estimated manually by experts based on copmplex rules.

## Download the data

In [ ]:
# Import important libraries for downloading the data.

import urllib.request
from pathlib import Path
import tarfile # to unzip the file
import pandas as pd

In [14]:
# Get the data

def get_hosing_data():
    tarball_path = Path("datasets/housing.tgz")
    if not tarball_path.is_file(): # if the file is not downloaded then do:
        Path("datasets").mkdir(parents=True, exist_ok=True)

        url = "https://github.com/ageron/data/raw/main/housing.tgz"
        urllib.request.urlretrieve(url, tarball_path)
    
    # Tar the file - unzip it
    with tarfile.open(tarball_path) as _:
            _.extractall(path="datasets")
    
    # Read the file
    return pd.read_csv("datasets/housing/housing.csv")

housing_full = get_hosing_data()

C:\Users\nesre\AppData\Local\Temp\ipykernel_11220\2139397772.py:13: DeprecationWarning: Python 3.14 will, by default, filter extracted tar archives and reject files or modify their metadata. Use the filter argument to control this behavior.
  _.extractall(path="datasets")


## A Quick Look at the Data

In [13]:
housing_full.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [16]:
housing_full.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


* All features are numerical, except ocean_proximity.
* Our dataset has 20,640 districts, i.e., rows (data samples).
* It contains 10 columns - features.
* total_rooms has missing values.

In [15]:
housing_full.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563400   
50%        435.000000   1166.000000    409.000000       3.534800   
75%        647.000000   1725.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

In [34]:
# Number of missing values
housing_full[housing_full["total_bedrooms"].isna()].shape[0]

207

* 207 missing values

In [35]:
# Thers in only one object feature - ocean_proximity

housing_full['ocean_proximity'].unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [37]:
housing_full["ocean_proximity"].value_counts()

ocean_proximity
<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: count, dtype: int64

* the colomn ocean_proximity has 5 unique values.

#### Data Visualization at glance

In [ ]:
import matplotlib.pyplot as plt

In [ ]:
housing_full.hist(bins=50, figsize=(12,8))

plt.show()

* Some distributions are Right skewed As a result some models may struggle to find patterns for such data distribution.

#### Create a Test Set

In [ ]:
from sklearn.model_selection import train_test_split
import numpy as np

* based on the convo with the client you found that income distribution really matters. So, it is better to use distribution based sampling instead of random sampling.

* If you split train/test randomly, some income groups may disappear or be under-represented.

In [ ]:
housing_full.head()

In [ ]:
housing_full["income_cat"] = pd.cut(housing_full["median_income"],
                                    bins=[0., 1.5, 3.0, 4.5, 6., np.inf],
                                    labels=[1, 2, 3, 4, 5])

In [ ]:
housing_full["income_cat"].head()

In [ ]:
cat_counts = housing_full["income_cat"].value_counts().sort_index()
cat_counts

In [ ]:
cat_counts.plot(kind="bar" ,rot=0, grid=True)

plt.xlabel("Income category")
plt.ylabel("Number of districts")
plt.show()

In [ ]:
housing_full.head()

In [ ]:
# Stratified Sampling.

strat_train_set, strat_test_set = train_test_split(housing_full, test_size=0.2, random_state=42,  stratify=housing_full["income_cat"])

In [ ]:
strat_train_set.shape

In [ ]:
strat_test_set.shape

In [ ]:
strat_test_set.shape[0]/housing_full.shape[0]

## Data Visualization to Gain Insights

In [ ]:
housing = strat_train_set.copy() # housing represents training data set.

### Visualize Geographical Data

In [ ]:
housing.head()

In [ ]:
housing.plot(kind="scatter", x="longitude", y="latitude", alpha=0.2, grid=True)

plt.show()

* It looks like California - the dataset is for houses in california, that is why!
* high population around the beach, and this makes sense.

In [ ]:
housing.plot(kind="scatter", x="longitude", y="latitude",
             s=housing["population"]/100, label="population",
             c="median_house_value", cmap="jet")

In [ ]:
# just graph compartion

housing.plot(kind="scatter", x="longitude", y="latitude", grid=True,
             s=housing["population"]/100, label="population",
             c="median_house_value", cmap="jet", colorbar=True,
             legend=True, sharex=False, figsize=(10,7))

### Correlations

In [ ]:
housing.corr(numeric_only=True)

In [ ]:
corr_matrix = housing.corr(numeric_only=True)["median_house_value"]
corr_matrix

In [ ]:
corr_matrix.sort_values(ascending=False)

* median income has a strong correlation

In [ ]:
from pandas.plotting import scatter_matrix

# lets go crazy

scatter_matrix(housing)

plt.show()

Oh, no!

In [ ]:
attributes = ["median_house_value", "median_income", "total_rooms", "housing_median_age"]

scatter_matrix(housing[attributes], figsize=(12,8))

plt.show()

In [ ]:
housing.plot(kind="scatter", x="median_income", y="median_house_value")

In [ ]:
housing.plot(kind="scatter", x="median_income", y="median_house_value", alpha=0.2, grid=True)

* there are are horizontal lines around 500k, 450k, 275k, 230k.....

## Prepare the Data for Machine Learning Algorithms

* lets separate the "predictors" and the "labels"

In [ ]:
housing = strat_train_set.drop("median_house_value", axis=1) # X_train
housing_labels = strat_train_set["median_house_value"].copy() # y_train

In [ ]:
housing.head(2)

In [ ]:
housing_labels.head()

### Clean Data

* Missing values

In [ ]:
from sklearn.impute import SimpleImputer

imputer = SimpleImputer(strategy="median")

In [ ]:
# only numerical values have median

housing_num = housing.select_dtypes(include="number")

In [ ]:
housing_num.head(2)

In [ ]:
imputer.fit(housing_num)

In [ ]:
housing_num.isna().sum()

In [ ]:
imputer.statistics_

In [ ]:
housing_num_filled = imputer.transform(housing_num)

In [ ]:
housing_num_filled

In [ ]:
housing_tr = pd.DataFrame(housing_num_filled, columns=housing_num.columns, index=housing_num.index)

In [ ]:
housing_tr.head(1) # Yo, it is pd df again!

### Handling Categorical Attributes

In [ ]:
housing_cat = housing[["ocean_proximity"]]
housing_cat.head()

In [ ]:
housing_cat["ocean_proximity"].unique()

In [ ]:
# Note: ordinal encoder is great for order based categories and it is not reocmmended of non ordered based categories like cities

from sklearn.preprocessing import OrdinalEncoder

ordinal_encoder = OrdinalEncoder()
housing_cat_encoded = ordinal_encoder.fit_transform(housing_cat)

In [ ]:
housing_cat_encoded

In [ ]:
ordinal_encoder.categories_

In [ ]:
from sklearn.preprocessing import OneHotEncoder

cat_encoder = OneHotEncoder()
housing_cat_1hot = cat_encoder.fit_transform(housing_cat)

In [ ]:
housing_cat_1hot

In [ ]:
#use .toarray() method to convert the sparse matrix to a NumPy array. note that NumPy array is heavier.

housing_cat_1hot.toarray()

Alternatively, you can set sparse_output=False when creating the OneHotEncoder (note: the sparse hyperparameter was renamned to sparse_output in Scikit-Learn 1.2)

In [ ]:
cat_encoder.categories_

### Feature Scaling and Transformation

In [ ]:
from sklearn.preprocessing import MinMaxScaler

min_max_scaler = MinMaxScaler(feature_range=(-1,1))
housing_num_min_max_scaled = min_max_scaler.fit_transform(housing_num)


In [ ]:
housing_num_min_max_scaled

In [ ]:
from sklearn.preprocessing import StandardScaler
std_scaler = StandardScaler()
housing_num_std_scaled = std_scaler.fit_transform(housing_num)

In [ ]:
housing_num_std_scaled

In [ ]:
from sklearn.linear_model import LinearRegression
target_scaler = StandardScaler()
scaled_labels = target_scaler.fit_transform(housing_labels.to_frame())

model = LinearRegression()
model.fit(housing[["median_income"]], scaled_labels)

some_new_data = housing[["median_income"]].iloc[:5]  # pretend this is new data
scaled_predictions = model.predict(some_new_data)

predictions = target_scaler.inverse_transform(scaled_predictions)

In [ ]:
from sklearn.compose import TransformedTargetRegressor

model = TransformedTargetRegressor(LinearRegression(),
transformer=StandardScaler())

model.fit(housing[["median_income"]], housing_labels)
predictions = model.predict(some_new_data)

### Transfomation Pipelines

In [ ]:
# a small pipeline for numerical attributes

from sklearn.pipeline import Pipeline

num_pipeline = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),
    ("standardized", StandardScaler()),
])

num_pipeline

In [ ]:
from sklearn.pipeline import make_pipeline
from sklearn.compose import make_column_selector, make_column_selector, make_column_transformer
from sklearn.impute import SimpleImputer

num_pipeline = make_pipeline(
    SimpleImputer(strategy="median"),
    StandardScaler()
)

cat_pipeline = make_pipeline(
    SimpleImputer(strategy="most_frequent"),
    OneHotEncoder(handle_unknown="ignore")
    #why handle_unknown?
)


preprocessing = make_column_transformer(
    (num_pipeline, make_column_selector(dtype_include=np.number)),
    (cat_pipeline, make_column_selector(dtype_include=object))
)

In [ ]:
housing_prepared = preprocessing.fit_transform(housing)
housing_prepared

In [ ]:
preprocessing.get_feature_names_out()